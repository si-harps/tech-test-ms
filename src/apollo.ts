import { ApolloClient, InMemoryCache, NormalizedCacheObject } from "@apollo/client";
import { useMemo } from "react";

let apolloClient: ApolloClient<NormalizedCacheObject>;

const createIsomorphicLink = () => {
  if (typeof window === "undefined") {
    // Serverside
  } else {
    const { HttpLink } = require("@apollo/client/link/http");
    return new HttpLink({
      uri: "http://localhost:3001/graphql"
    });
  }
}

const createApolloClient = () => new ApolloClient({
  ssrMode: typeof window === "undefined",
  link: createIsomorphicLink(),
  cache: new InMemoryCache()
})

export const initializeApollo = (initialState = null) => {

  const _apolloClient = apolloClient ?? createApolloClient();
  
  if (initialState) {
    _apolloClient.cache.restore(initialState);
  }

  if (typeof window === "undefined")
    return _apolloClient;

   apolloClient = apolloClient ?? _apolloClient;

   return apolloClient;
}

export const useApollo = (initialState: object) => {
  const store = useMemo( () => initializeApollo(initialState), [initialState]);
  return store;
}