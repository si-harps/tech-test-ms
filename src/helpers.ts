export const priceConversion = (digits: number): string => {
  return (digits/100).toFixed(2);
}