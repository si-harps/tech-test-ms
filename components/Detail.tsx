import styles from "./Detail.module.css";
import classnames from "classnames";
import { priceConversion } from "../src/helpers";

interface Props {
  id: number;
  name: string;
  image: string;
  currentPrice: number;
  originalPrice?: number;
  text: string;
}

const Detail: React.FC<Props> = ({ id, name, image, currentPrice, originalPrice, text }) => {
  return <div key={id} className={styles.detail}>
    <h3 className={styles.title}>{ name }</h3>
    <img className={styles.asset} src={`https://asset1.cxnmarksandspencer.com/is/image/mands/${image}`} />
    <span className={styles.row}>£{ priceConversion(currentPrice) }</span>
    { originalPrice && <span className={classnames(styles.row, styles.strikethrough)}>
      Original Price: £{ priceConversion(originalPrice) }
    </span> }
    <p className={styles.description}>{ text }</p>
  </div>
}

export default Detail;