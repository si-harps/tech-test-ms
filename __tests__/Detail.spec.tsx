import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import Detail from "../components/Detail";
import { priceConversion } from "../src/helpers";

describe('Detail', () => {

  let expectedProps;

  beforeEach( () => {
    expectedProps = {
      id: "2",
      name: "Shirt",
      currentPrice: 1000,
      originalPrice: 1500,
      text: 'Lorem ipsum'
    }
  })

  test('should render name, current price and original price', () => {

    const { getByText } = render(<Detail {...expectedProps} />);
    const name = getByText(expectedProps.name);
    const currentPrice = getByText(`£${priceConversion(expectedProps.currentPrice)}`);
    const originalPrice = getByText(`Original Price: £${priceConversion(expectedProps.originalPrice)}`);
    const text = getByText(expectedProps.text);

    expect(name).toBeVisible();
    expect(currentPrice).toBeVisible();
    expect(originalPrice).toBeVisible();
    expect(text).toBeVisible();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<Detail {...expectedProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
})