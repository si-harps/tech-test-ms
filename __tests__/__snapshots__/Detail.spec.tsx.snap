// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Shirt
    </h3>
    <img
      src="https://asset1.cxnmarksandspencer.com/is/image/mands/undefined"
    />
    <span>
      £10.00
    </span>
    <span
      class=""
    >
      Original Price: £15.00
    </span>
    <p>
      Lorem ipsum
    </p>
  </div>
</DocumentFragment>
`;
